rfc/src/rfcs/0008-function-contracts.md

```diff
@@ -0,0 +1,580 @@
+
+```rs
+fn my_div(dividend: u32, divisor: u32) -> u32 {
+  dividend / divisor
+}
+```
+
+1. In the first phase we **specify** the approximation. Kani provides two new
+   annotations: `requires` (preconditions) to describe the expectations this
+   function has as to the calling context and `ensures` (postconditions) which
+   approximates function outputs in terms of function inputs.
+
+   ```rs
+   #[kani::requires(divisor != 0)]
+   #[kani::ensures(result <= dividend)]
+   fn my_div(dividend: u32, divisor: u32) -> u32 {
+     dividend / divisor
+   }
+   ```
+  
+   `requires` here indicates this function expects its `divisor` input to never
+   be 0, or it will not execute correctly (i.e. panic).
+
+   `ensures` puts a bound on the output, relative to the `dividend` input.
+
+   Conditions in contracts are plain Rust expressions which can reference the
+   function arguments and, in case of `ensures`, the result of the function as a
+   special `result`[^result-naming] varaible. Syntactically Kani supports any
+   Rust expression, including function calls, defining types etc, however they
+   must be side-effect free[^side-effects]. Multiple `requires` and `ensures`
+   clauses are allowed on the same function, they are implicitly logically
+   conjoined.
+
+   [^result-naming]: See [open questions](#open-questions) for a discussion
+       about naming of the result variable.
+
+2. Checking the Contract
+
+   Next Kani must make sure that the approximation we specified actually holds.
+   This is in opposition to the related "stubbing" mechanism, where the
+   approximation is not checked but always trusted.
+
+   The contract must always overapproximate the function to guarantee soundness.
+   Specifically the domain (inputs) of the function described (by the `requires`
+   clause) must be *at most* as large as the input space for which the
+   function's behavior is well defined, and the codomain (outputs, described by
+   `ensures`) must be *at least* as large as the actual space of outputs the
+   function may produce.
+
+   For example in our case it would be permissible to use
+   `#[kani::requires(divisor > 100)]` (smaller permissible input domain) or
+   `#[kani::ensures(result < dividend + divisor)]` (larger possible output
+   domain), but `#[kani::ensures(result < dividend)]` (small output) is not
+   allowed.
+
+   To facilitate the check Kani needs a suitable environment to verify the
+   function in. For this proposal the environment must be provided by us (the
+   users). See [future possibilities](#future-possibilities) for a discussion
+   about the arising soundness issues and their remedies.
+
+   We provide the checking environment for our contract with a special new
+   `proof_for_contract` harness.
+
+   ```rs
+   #[kani::proof_for_contract(my_div)]
+   fn my_div_harness() {
+     my_div(kani::any(), kani::any())
+   }
+   ```
+
+   Similar to a unit-test harness for any other function, we are supposed to
```

Comment:
Maybe just mention that users must ensure that these harnesses provide all possible input combinations to the function. I think this is more relevant when handling references.

Also, maybe say "verification harness" instead of 'unit-test harness". I read the latter as concrete tests, not symbolic.